File: src/main/java/com/github/stars/controller/AuthorController.java
package com.github.stars.controller;

import com.baomidou.mybatisplus.core.metadata.IPage;
import com.github.stars.dto.AuthorDTO;
import com.github.stars.entity.GithubRepo;
import com.github.stars.service.AuthorService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.annotation.Resource;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
public class AuthorController {

    @Resource
    private AuthorService authorService;

    /**
     * 作者列表 JSON API（分页 + 搜索）
     */
    @GetMapping("/api/authors")
    public Map<String, Object> apiAuthors(@RequestParam(value = "page", defaultValue = "1") int page,
                                          @RequestParam(value = "size", defaultValue = "24") int size,
                                          @RequestParam(value = "keyword", defaultValue = "") String keyword) {

        IPage<AuthorDTO> pageResult = authorService.getAuthorPage(page, size, keyword);

        Map<String, Object> result = new HashMap<>();
        result.put("records", pageResult.getRecords());
        result.put("total", pageResult.getTotal());
        result.put("size", pageResult.getSize());
        result.put("current", pageResult.getCurrent());
        result.put("pages", pageResult.getPages());
        return result;
    }

    /**
     * 作者仓库列表 JSON API（分页 + 排序）
     */
    @GetMapping("/api/authors/{ownerName}")
    public Map<String, Object> apiAuthorRepos(@PathVariable("ownerName") String ownerName,
                                              @RequestParam(value = "page", defaultValue = "1") int page,
                                              @RequestParam(value = "size", defaultValue = "12") int size,
                                              @RequestParam(value = "sortBy", defaultValue = "starred_at") String sortBy,
                                              @RequestParam(value = "sortOrder", defaultValue = "desc") String sortOrder) {

        IPage<GithubRepo> pageResult = authorService.getAuthorRepos(ownerName, page, size, sortBy, sortOrder);

        Map<String, Object> result = new HashMap<>();
        result.put("records", pageResult.getRecords());
        result.put("total", pageResult.getTotal());
        result.put("size", pageResult.getSize());
        result.put("current", pageResult.getCurrent());
        result.put("pages", pageResult.getPages());
        return result;
    }

    /**
     * 导出某作者的所有仓库链接为 txt 文件
     */
    @GetMapping("/api/authors/{ownerName}/export")
    public ResponseEntity<byte[]> exportAuthorUrls(@PathVariable("ownerName") String ownerName,
                                                    @RequestParam(value = "sortBy", defaultValue = "starred_at") String sortBy,
                                                    @RequestParam(value = "sortOrder", defaultValue = "desc") String sortOrder) {

        List<String> urls = authorService.getAuthorAllRepoUrls(ownerName, sortBy, sortOrder);
        String content = String.join("\n", urls);
        byte[] bytes = content.getBytes(java.nio.charset.StandardCharsets.UTF_8);

        String filename = "github_stars_" + ownerName + ".txt";

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=" + filename)
                .contentType(MediaType.TEXT_PLAIN)
                .contentLength(bytes.length)
                .body(bytes);
    }
}

File: backend-nestjs/src/author/author.controller.ts
import { Controller, Get, Logger, Param, Query, Res } from '@nestjs/common';
import type { Response } from 'express';
import { AuthorService } from './author.service';

@Controller('api/authors')
export class AuthorController {
    private readonly logger = new Logger(AuthorController.name);

    constructor(private readonly service: AuthorService) {}

    /**
     * 分页获取作者列表
     *
     * 支持关键字搜索，按总 Star 数降序排列
     *
     * @param q.page 页码，默认1
     * @param q.size 每页条数，默认24
     * @param q.keyword 搜索关键字，按作者名模糊匹配
     * @returns 分页后的作者列表
     */
    @Get()
    async list(@Query() q: any) {
        return this.service.getAuthorPage(parseInt(q.page) || 1, parseInt(q.size) || 24, q.keyword || '');
    }

    /**
     * 获取指定作者的所有 Star 仓库
     *
     * 支持多种排序方式
     *
     * @param ownerName 作者名
     * @param q.page 页码，默认1
     * @param q.size 每页条数，默认12
     * @param q.sortBy 排序字段，默认 starred_at
     * @param q.sortOrder 排序方向，默认 desc
     * @returns 分页后的仓库列表
     */
    @Get(':ownerName')
    async repos(@Param('ownerName') owner: string, @Query() q: any) {
        return this.service.getAuthorRepos(
            owner,
            parseInt(q.page) || 1,
            parseInt(q.size) || 12,
            q.sortBy || 'starred_at',
            q.sortOrder || 'desc',
        );
    }

    /**
     * 导出指定作者的所有 Star 仓库 URL
     *
     * 以纯文本文件下载，每行一个 GitHub 仓库地址
     *
     * @param ownerName 作者名
     * @param q.sortBy 排序字段，默认 starred_at
     * @param q.sortOrder 排序方向，默认 desc
     * @param res Express Response 对象，用于设置下载头并返回文件内容
     */
    @Get(':ownerName/export')
    async export(@Param('ownerName') owner: string, @Query() q: any, @Res() res: Response) {
        const urls = await this.service.getAuthorAllRepoUrls(owner, q.sortBy || 'starred_at', q.sortOrder || 'desc');
        res.set({
            'Content-Type': 'text/plain; charset=utf-8',
            'Content-Disposition': `attachment; filename*=UTF-8''${encodeURIComponent(owner + '-stars.txt')}`,
        });
        res.send(urls.join('\n'));
    }
}
